File: 2018-minimum-space-wasted-from-packaging/minimum-space-wasted-from-packaging.php
class Solution {

    function minWastedSpace($packages, $boxes) {
        $MOD = 1000000007;

        sort($packages);

        $n = count($packages);

        $prefix = array_fill(0, $n + 1, 0);

        for ($i = 0; $i < $n; $i++) {
            $prefix[$i + 1] = $prefix[$i] + $packages[$i];
        }

        $ans = PHP_INT_MAX;

        foreach ($boxes as $supplier) {
            sort($supplier);

            if ($supplier[count($supplier) - 1] < $packages[$n - 1]) {
                continue;
            }

            $waste = 0;
            $idx = 0;

            foreach ($supplier as $b) {
                $next = $this->upperBound($packages, $idx, $b);

                $waste += ($next - $idx) * $b - ($prefix[$next] - $prefix[$idx]);

                $idx = $next;

                if ($idx == $n) {
                    break;
                }
            }

            $ans = min($ans, $waste);
        }

        return $ans == PHP_INT_MAX ? -1 : $ans % $MOD;
    }

    private function upperBound($arr, $start, $target) {
        $l = $start;
        $r = count($arr);

        while ($l < $r) {
            $mid = intdiv($l + $r, 2);

            if ($arr[$mid] <= $target) {
                $l = $mid + 1;
            } else {
                $r = $mid;
            }
        }

        return $l;
    }
}

File: 2018-minimum-space-wasted-from-packaging/minimum-space-wasted-from-packaging.cpp
class Solution {
public:
    int minWastedSpace(vector<int>& packages, vector<vector<int>>& boxes) {
        const int MOD = 1e9 + 7;

        sort(packages.begin(), packages.end());

        int n = packages.size();

        vector<long long> prefix(n + 1, 0);
        for (int i = 0; i < n; i++)
            prefix[i + 1] = prefix[i] + packages[i];

        long long ans = LLONG_MAX;

        for (auto &supplier : boxes) {
            sort(supplier.begin(), supplier.end());

            if (supplier.back() < packages.back())
                continue;

            long long waste = 0;
            int idx = 0;

            for (int b : supplier) {
                int next = upper_bound(packages.begin() + idx, packages.end(), b) - packages.begin();

                long long sum = prefix[next] - prefix[idx];
                waste += 1LL * (next - idx) * b - sum;

                idx = next;
                if (idx == n) break;
            }

            ans = min(ans, waste);
        }

        return ans == LLONG_MAX ? -1 : ans % MOD;
    }
};